Add specific exception message when Explodes cannot find a weapon

reaperrr opened the pull request:
Additionally, accept empty string to 'disable' `Explodes` weapons.

Last but not least, since we define `UnitExplode` as default for both filled and empty weapon, the d2k grenadier used it when reloading(!), causing massive damage.

Adresses #8940 (or at least helps debugging that crash).

reaperrr committed:
Add Explodes safeguards

Accept both null and empty string to disable explosion weapons.

Throw specific exception if weapon was not found.
Require Weapon (EmptyWeapon stays optional).

reaperrr commented:
Note: I'm not sure about defaulting to 'null' (3rd commit). Generally speaking I think this is better because those magic, not-visible-in-yaml defaults are always troublesome for modders.

On the other hand, it slightly increases the risk of regressions in the shipping mods, unless we catch any existing problems before this is merged.

pchote commented on OpenRA.Mods.Common/Traits/Explodes.cs:
```diff
@@ -32,14 +33,20 @@ public class ExplodesInfo : ITraitInfo
+		{
+			this.info = info;
+
+			if (info.Weapon == null && info.EmptyWeapon == null)
```
Could we use `[Require]` on the traitinfo instead?  I don't think it would be that big of a deal to force modders to define both the empty and loaded versions, and that gives us static guarantees.

pchote replied:
Or we could `[Require]` only the `Weapon`, and leave `EmptyWeapon` optional and defaulting to null.

reaperrr commented:
Updated.

pchote commented on OpenRA.Mods.Common/Traits/Explodes.cs:
```diff
@@ -54,9 +55,13 @@ public void Killed(Actor self, AttackInfo e)
+			if (string.IsNullOrEmpty(weaponName))
 				return;
 
+			if (!e.Attacker.World.Map.Rules.Weapons.ContainsKey(weaponName.ToLowerInvariant()))
```
The `WeaponReference` already gives us this, so i'm not convinced we need this.  I guess this is going against my comments in #8940 however, so feel free to ignore me.

abcdefg30 commented on OpenRA.Mods.Common/Traits/Explodes.cs:
```diff
@@ -17,11 +18,11 @@ namespace OpenRA.Mods.Common.Traits
 
 		[WeaponReference, Desc("Weapon to use for explosion if no ammo/payload is loaded.")]
-		public readonly string EmptyWeapon = "UnitExplode";
+		public readonly string EmptyWeapon = null;
```
Are we sure that this change doesn't want an upgrade rule?

abcdefg30 replied:
There are several places in our default mods yaml where we don't explicitly give an `EmptyWeapon`.

abcdefg30 commented:
Looks good to me otherwise. :+1: / :white_check_mark:

reaperrr committed:
Fixed D2k grenadier to not explode violently when reloading

Removed redundant Chance (default is 100 anyway).

reaperrr commented:
Currently on't have the time and 'head-bandwidth' to deal with an upgrade rule, so I reverted the null-default for EmptyWeapon and fixed the d2k grenadier manually instead.

penev92 commented:
:+1:
